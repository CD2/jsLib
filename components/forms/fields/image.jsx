import React from 'react'
import PropTypes from 'prop-types'

import { observable, action, computed } from 'mobx'
import { observer } from 'mobx-react'

@observer
export default class ImageField extends React.Component {

  @observable preview_src = ``

  static propTypes = {
    disabled: PropTypes.bool,
    multiple: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onRawChange: PropTypes.func,
  }

  handleChange = (e) => {
    const { onRawChange, onChange } = this.props
    if (onRawChange) onRawChange(e)
    if (onChange) {onChange({
      name: e.target.name,
      value: e.target.files[0],
      filename: e.target.files[0].name,
      size: e.target.files[`0`].size,
      type: e.target.files[`0`].type,
    })}
  }

  componentWillReceiveProps(props) {
    if (!props.value) return
    if (props.value instanceof File) {
      var reader = new FileReader()
      reader.onload = action((e) => this.preview_src = e.target.result)
      reader.readAsDataURL(props.value)
    } else {
      this.preview_src = props.value
    }
  }

  renderPreview() {
    const { value } = this.props
    if (value instanceof File) {
    }
    return (
      <div>
        <img src={this.preview_src} />
      </div>
    )
  }

  render() {
    const { name, multiple, onFocus } = this.props

    return (
      <div>
        {this.renderPreview()}
        <input
          type="file"
          name={name}
          multiple={multiple}
          onChange={this.handleChange}
          onFocus={onFocus}
          disabled={this.props.disabled}
        />
      </div>
    )
  }

}
