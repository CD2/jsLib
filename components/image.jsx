import React from 'react'
import PropTypes from 'prop-types'
import { buildUrl } from 'utils/api_http'
import invariant from 'invariant'
import { styled } from 'utils/theme'
import decorate from 'utils/decorate'

export class Image extends React.Component {

  static propTypes = {
    alt: PropTypes.string,
    background: PropTypes.bool,
    children: PropTypes.node,
    circular: PropTypes.bool,
    className: PropTypes.string,
    crop: PropTypes.bool,
    defaultSrc: PropTypes.string,
    height: PropTypes.number,
    onClick: PropTypes.func,
    size: PropTypes.string,
    style: PropTypes.object,
    uid: PropTypes.string,
    width: PropTypes.number,
  }

  static defaultProps = {
    crop: false,
    background: false,
    alt: ``,
  }

  get url() {
    const { uid, width, height, crop, size } = this.props
    const params = { uid, size }
    if (height && !size) params.size = `${Math.round(height*1.5*2)}x${height*2}`
    if (width && !size) params.size = `${width*2}x${Math.round(width/1.5*2)}`
    if (width && height && !size) params.size = `${width*2}x${height*2}`
    if (crop) params.crop = true
    return buildUrl([`/image`], params)
  }

  render() {
    const { alt, background, children, defaultSrc, uid } = this.props

    invariant(!(background && alt), `background images don't accept alt tags`)

    let url = this.url
    if (!uid) { url = defaultSrc }

    if (background) {
      return (
        <div
          style={
            Object.assign({ backgroundImage: `url(${url})` }, this.props.style)
          }
          children={children}
          className={`background-image ${this.props.className}`}
          onClick={this.props.onClick}
        />
      )
    }

    invariant(!children, `images cant accept children unless marked as a background image`)

    return (
      <img
        alt={alt}
        src={url}
        className={`image ${this.props.className}`}
        style={this.props.style}
        onClick={this.props.onClick}
      />
    )
  }

}
export default decorate(
  styled`
    background-size: cover;
    background-position: 50%;
    background-repeat: no-repeat;
    position: relative;

    ${({ circular }) => {
    if (circular) {
      return `border-radius: 50%;
      `
    }
  }}
    ${({ height }) => {
    if (height) {
      return `height: ${height}px;`
    }
  }}
    ${({ width }) => {
    if (width) {
      return `width: ${width}px;`
    }
  }}

  `,
  Image
)
