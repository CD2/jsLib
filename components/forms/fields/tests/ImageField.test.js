/* eslint-disable react/jsx-filename-extension */
import React from 'react'
import { ImageField } from '../image'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe(`<ImageField />`, () => {
  const view = shallow((
    <ImageField />
  ))

  it(`snapshot`, ()=>expect(toJson(view)).toMatchSnapshot())
})