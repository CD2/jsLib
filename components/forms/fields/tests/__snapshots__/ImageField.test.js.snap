// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageField /> snapshot 1`] = `
<div>
  <div />
  <input
    onChange={[Function]}
    type="file"
  />
</div>
`;
